commerce/domain/promotion/cart/calculator/classes/PromotionCalculatorSample.cls

```diff
@@ -1,24 +1,53 @@
- /**
+/**
   * @description This sample is for the situations where Promotion Calculation needs to be extended or overridden via the
   * extension point for the Promotion Calculator. You are expected to refer this and write your own implementation.
   * This class must extend the CartExtension.PromotionsCartCalculator class to be processed.
+  *
+  * In this example cart items are evaluated against a BOGO (Buy X, Get Y) promotion > Buy 5 items of qualifying product X,
+  * get $2 off a unit of target Y.
   */
 public with sharing class PromotionCalculatorSample extends CartExtension.PromotionsCartCalculator {
 
-    // You MUST change this to be a valid promotion id.
-    public static final String DUMMY_PROMOTION_ID = '0c8xx000000003FAAQ';
+    // You MUST change following to be a valid promotion, product ids.
+    public static final String DUMMY_PROMOTION_ID = '0c8xx00000004JlAAI';
+    private static final String QUALIFIER_PRODUCT_ID = '01txx0000006lmmAAA';
+    private static final String TARGET_PRODUCT_ID = '01txx0000006lmuAAA';
+    private static final Integer PROMOTION_ADJUSTMENT = -2;
+    private static final Integer QUALIFIER_QUANTITY = 5;
 
     public virtual override void calculate(CartExtension.CartCalculateCalculatorRequest request) {
-        cartextension.Cart cart = request.getCart();
-        resetAllAdjustments(cart);
-        applyAdjustments(cart);
+        validateBuyerActionDetailsAndEvaluatePromotion(request.getCart(), request.getOptionalBuyerActionDetails());
+    }
+
+	/**
+	 * @description Evaluate promotion for the cart items if qualifier and/or target products part of changed elements in cart.
+	 * @param cart Holds details about cart
+	 * @param optionalBuyerActionDetails The latest set of changes applied to the Cart by the Buyer
+	 */
+    private void validateBuyerActionDetailsAndEvaluatePromotion(cartextension.Cart cart, cartextension.OptionalBuyerActionDetails optionalBuyerActionDetails) {
+        if (optionalBuyerActionDetails.isPresent() || !optionalBuyerActionDetails.get().isCheckoutStarted()) {
+            List<CartExtension.CartItemChange> cartItemChanges = optionalBuyerActionDetails.get().getCartItemChanges();
+
+            for (CartExtension.CartItemChange cartItemChange : cartItemChanges) {
+                 CartExtension.OptionalCartItem optionalCartItem = cartItemChange.getChangedItem();
+                 if (optionalCartItem.isPresent()) {
+                     CartExtension.CartItem cartItem = optionalCartItem.get();
+                     if (cartItem.getProduct2Id() == ID.valueOf(QUALIFIER_PRODUCT_ID) ||
+                         cartItem.getProduct2Id() == ID.valueOf(TARGET_PRODUCT_ID)) {
+                     		resetAllAdjustments(cart);
+                        	evaluatePromotionForCartItems(cart);
+                        	break;
```

Comment:
Do we want to break here? what if there are more than one matching items?

Reply:
Ignore, i guess for this example its ok.

Reply:
If buyer added a qualifying or target product we recalculate the whole cart. So, it's ok to break.